lib/Bio/EnsEMBL/DataCheck/Test/DataCheck.pm

```diff
@@ -278,16 +278,33 @@ sub row_subtotals {
   $sql2 = $sql1 if ! defined $sql2;
   $min_proportion = 1 if ! defined $min_proportion;
 
-  if ($sql1 !~ /^\s*SELECT\s+[^,]+\s*,\s*COUNT[^,]+FROM.+GROUP\s+BY/ms) {
-    die "Invalid SQL statement for subtotals. Must select a single column first, then a count.\n($sql1)";
-  }
-  if ($sql2 !~ /^\s*SELECT\s+[^,]+\s*,\s*COUNT[^,]+FROM.+GROUP\s+BY/ms) {
-    die "Invalid SQL statement for subtotals. Must select a single column first, then a count.\n($sql2)";
-  }
-
   my ( undef, $rows1 ) = _query( $dbc1, $sql1 );
   my ( undef, $rows2 ) = _query( $dbc2, $sql2 );
 
+  my $len1 = @$rows1;
+  my $len2 = @$rows2;
+
+  if ($len1 > 0) {
+    my $len_elem1 = @{@$rows1[0]};
+    if ($len_elem1 != 2) {
+      die "Invalid SQL query for row_subtotals. Must return exactly two columns, a key and a number.\n($sql1)"
+    }
+    else {
+      my $elem1 = @$rows1[0];
+      die "Invalid SQL query for row_subtotal. Second column must be a number.\n($sql1)" unless @$elem1[1] =~ m/[0-9]+/;
+    }
+  }
+  if ($len2 > 0) {
+    my $len_elem2 = @{@$rows2[0]};
+    if ($len_elem2 != 2) {
+      die "Invalid SQL query for row_subtotals. Must return exactly two columns, a key and a number.\n($sql2)"
+    }
+    else {
+      my $elem2 = @$rows2[0];
+      die "Invalid SQL query for row_subtotals. Second column must be a number.\n($sql2)" unless @$elem2[1] =~ m/[0-9]+/;
+    }
+  }
+
```

Comment:
@james-monkeyshines There's probably is a better way to check length and type of the query returned elements, but my knowledge of Perl is quite limited so I'm very open to suggestions 🙂

Reply:
This approach is fine, treating the array as a scalar to get the count is the Perlish way to do it (and this is much nicer than the former parsing of the SQL query). But the dereferencing of the array elements is maybe a little non-standard (although it works fine), so I'll make a couple of suggestions there...